narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,201 @@
+                .expect("VALIDITY_THRESHOLD should fit in u16"),
+            protocol_config.random_beacon_reduction_allowed_delta(),
+        );
+        let party = match dkg::Party::<PkG, EncG>::new(
+            private_key,
+            nodes,
+            t.into(),
+            fastcrypto_tbls::random_oracle::RandomOracle::new(
+                format!("dkg {}", committee.epoch()).as_str(),
+            ),
+            &mut rand::thread_rng(),
+        ) {
+            Ok(party) => party,
+            Err(err) => {
+                error!("Error while initializing random beacon Party: {err:?}");
+                return None;
+            }
+        };
+        info!("random beacon: state initialized");
+        Some(Self {
+            party,
+            messages: Vec::new(),
+            processed_messages: Vec::new(),
+            shares: dkg::SharesMap::with_capacity(0),
+            confirmations: Vec::new(),
+            dkg_output: None,
+        })
+    }
+
+    async fn start_dkg(&self, tx_system_messages: &Sender<SystemMessage>) {
+        let msg = self.party.create_message(&mut rand::thread_rng());
+        info!("random beacon: sending DKG Message: {msg:?}");
+        let _ = tx_system_messages
+            .send(SystemMessage::DkgMessage(msg))
+            .await;
+    }
+
+    fn add_message(&mut self, msg: dkg::Message<PkG, EncG>) {
+        if !self.shares.is_empty() {
+            // We've already sent a `Confirmation`, so we can't add any more messages.
+            return;
+        }
+        self.messages.push(msg.clone());
+        match self.party.process_message(msg, &mut rand::thread_rng()) {
+            Ok(processed) => {
+                self.processed_messages.push(processed);
+            }
+            Err(err) => {
+                debug!("error while processing randomness DKG message: {err:?}");
+            }
+        }
+    }
+
+    fn add_confirmation(&mut self, conf: dkg::Confirmation<EncG>) {
+        self.confirmations.push(conf)
+    }
+
+    // Generates the next SystemMessage needed to advance the random beacon protocol, if possible,
+    // and sends it to the proposer.
+    async fn advance(&mut self, tx_system_messages: &Sender<SystemMessage>) {
+        // Once we have enough ProcessedMessages, send a Confirmation.
+        if self.shares.is_empty() && !self.processed_messages.is_empty() {
+            match self.party.merge(&self.processed_messages) {
+                Ok((shares, conf)) => {
+                    info!(
+                        "random beacon: sending DKG Confirmation with {} complaints",
+                        conf.complaints.len()
+                    );
+                    self.shares = shares;
+                    let _ = tx_system_messages
+                        .send(SystemMessage::DkgConfirmation(conf))
+                        .await;
+                }
+                Err(fastcrypto::error::FastCryptoError::InputTooShort(_)) => (), // wait for more input
```

Comment:
This error is a bit too generic (might indicate other "input too short" error from an internal function). Let's add a TODO to change this error to a more specific one (I'll do that next week)

Reply:
Done